### PR TITLE
SDK Access Boundaires

### LINKED ISSUE #255: [Feature Contribution]: Create access boundaries in the SDK
### What happened?

Most models, helpers, classes, and interfaces in the Carbon Aware SDK are currently public. With the addition of a library nuget package, this means that now anyone who downloads the package has access to all these functions and can call them. This increases confusion and leaves open unintended use cases or not-supported paths.

In order to better support the correct use of the SDK, we will add access boundaries by making most public things internal. Projects that need access to the internal variables are given access selectively

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Feature Commitment

- [X] I commit to contributing this feature as a PR and working with the GSF to merge this feature into the Carbon Aware SDK.

### PR DESCRIPTION
**Pull request by JenMadiedo**
Issue Number: #255 

## Summary
Make most non-client facing classes internal. See [ExcaliDraw](https://excalidraw.com/#json=3B5FsCS3qdlZ_RRRjrDJ_,4MGXf7ZzZgLTHfMADRceBw) drawing for visualization of the internal read access for all the major projects in the SDK. 

## Changes

- Make most models, interfaces, and classes internal
- Use `InternalsVisibleTo` tag in .csproj files to selectively give access to projects that may need to see certain internals
- Remove an unused imports given new restrictions

## Checklist

- [X] Local Tests Passing?
- [X] CICD and Pipeline Tests Passing?
- [X] Added any new Tests?
- [ ] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [x] This is not a breaking change. If it is, please describe it below.

This PR Closes #255

### CODE REVIEW

**Review by gfmatthews — CHANGES_REQUESTED** [review]
This is a big PR and it's going to really help the maintainability of this repo over the long term! Nice work! A couple questions and comments in line.

Additionally a question - have we done some sanity checking to make sure that the WebAPI and other known endpoints continue to function correctly? We have some customers out there using this code and I'd want to do a double-check that we aren't introducing any known regressions with this change (especially given the scope)

File: src/CarbonAware.DataSources/CarbonAware.DataSources.Json/src/CarbonAware.DataSources.Json.csproj [hunk 1]
```diff
@@ -17,4 +12,15 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>
   </ItemGroup>
+
+	<ItemGroup>
+    <InternalsVisibleTo Include="CarbonAware.DataSources.Json.Mocks" />
+		<InternalsVisibleTo Include="CarbonAware.DataSources.Json.Tests" />
```
Comment (juzuluag) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.Json/src/CarbonAware.DataSources.Json.csproj:18]: nit: Indentation

File: src/CarbonAware/src/CarbonAware.csproj [hunk 2]
```diff
@@ -16,4 +16,33 @@
     <PackageReference Include="System.ComponentModel.Composition"/>
   </ItemGroup>
 
+  <ItemGroup>
+    <InternalsVisibleTo Include="CarbonAware.Aggregators"/>
+	<InternalsVisibleTo Include="CarbonAware.Aggregators.Tests"/>
+    <InternalsVisibleTo Include="caw"/>
+	<InternalsVisibleTo Include="CarbonAware.CLI.IntegrationTests"/>
```
Comment (juzuluag) [inline src/CarbonAware/src/CarbonAware.csproj:23]: nit: indentation

**Review by juzuluag — COMMENTED** [review]
Thanks for doing this 👍 . Just a few comments to consider.

File: src/CarbonAware.Aggregators/test/CarbonAwareParametersTests.cs [hunk 3]
```diff
@@ -6,12 +6,12 @@
...
 using System.Linq;
 using System.Text.Json.Serialization;
-using PropertyName = CarbonAware.Aggregators.CarbonAware.CarbonAwareParameters.PropertyName;
+using static CarbonAware.Aggregators.CarbonAware.CarbonAwareParameters;
 
 namespace CarbonAware.Aggregators.Tests;
 
 [TestFixture]
-public class CarbonAwareParametersTests
+class CarbonAwareParametersTests
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.Aggregators/test/CarbonAwareParametersTests.cs:14]: Making these default let us also make all the test bases and types internal. It's just for tests but also still keeps the circle small

File: src/CarbonAware.Aggregators/src/Configuration/ServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 namespace CarbonAware.Aggregators.Configuration;
 
-public static class ServiceCollectionExtensions
+internal static class ServiceCollectionExtensions
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.Aggregators/src/Configuration/ServiceCollectionExtensions.cs:14]: Fixed, thanks for the catch

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/Configuration/ServiceCollectionExtensions.cs [hunk 5]
```diff
@@ -1,4 +1,5 @@
 using CarbonAware.Configuration;
+using CarbonAware.Interfaces;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/Configuration/ServiceCollectionExtensions.cs]: Fixed, thanks for the catch

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs [hunk 6]
```diff
@@ -1,18 +1,18 @@
-﻿using CarbonAware.DataSources.WattTime.Client;
+﻿using CarbonAware.Interfaces;
+using CarbonAware.DataSources.WattTime.Client;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs]: Fixed, thanks for the catch

File: src/CarbonAware.WebApi/src/Program.cs [hunk 7]
```diff
@@ -48,8 +49,17 @@
 
 builder.Services.Configure<CarbonAwareVariablesConfiguration>(builder.Configuration.GetSection(CarbonAwareVariablesConfiguration.Key));
 
-string? errorMessage;
-bool successfulEmissionServices = builder.Services.TryAddCarbonAwareEmissionServices(builder.Configuration, out errorMessage);
+bool successfulServices = true;
```
[the remark this answers is not in the document]
Comment (JenMadiedo) [reply src/CarbonAware.WebApi/src/Program.cs:52]: I agree, I would take this comment to https://github.com/Green-Software-Foundation/carbon-aware-sdk/pull/245 as that's from that PR's commit. This PR inherits from that one (and the one removing the aggregator from the library) so you need to individually select the access boundary commits or use the link in the summary

**Review by gfmatthews — APPROVED** [review]
Thanks for the changes! Looks great!

**Review by juzuluag — APPROVED** [review]
Great 👍

**Comment by JenMadiedo** [comment]
@vaughanknight this is ready for review

**Review by vaughanknight — APPROVED** [review]
LGTM.  Big coss sweeping change with a lot of effort in it.  Really appreciate the contribution.

**Pull request merged**